### PR TITLE
Analyses Listing

### PR DESCRIPTION
**Pull request by sri2k1us**
- Analyses Listing
- Ability to filter by `App Type` or `Permission`
- Click on Analysis name to go to output folder
- Click on Interactive link to go to VICE analyses
- Click on HT icon to drill down to HT analyses details

- Basic Listing
image

- HT drill down
image

- iPhone SE - HT Drill down view
image

### CODE REVIEW

File: src/components/analyses/AnalysesNavigation.js [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    },
+}));
+
+function getOwnershipFilters() {
+    return Object.keys(ownershipFilter).map((filter) => {
+        return {
+            name: ownershipFilter[filter],
+        };
+    });
+}
```
Comment (ash3rz) [inline src/components/analyses/AnalysesNavigation.js:62]: I think you could also do: 

```suggestion
function getOwnershipFilters() {
    return Object.values(ownershipFilter).map((filter) => {
        return {
            name: filter,
        };
    });
}
```

File: src/components/analyses/AnalysesNavigation.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            </Tooltip>
+            <Hidden xsDown>
+                <Typography variant="caption" className={classes.batchFilter}>
+                    {getMessage("viewingBatch", {
+                        values: { name: name },
+                    })}
+                </Typography>
+
+                <Divider
+                    style={{ margin: 8 }}
```
Comment (ash3rz) [inline src/components/analyses/AnalysesNavigation.js:84]: This should probably be one of the `theme.spacing` values?

File: src/components/analyses/Header.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    toolbar: {
+        backgroundColor: theme.palette.primary.main,
+    },
+    menuButton: {
+        color: theme.palette.primary.contrastText,
+    },
+}));
+
+function getOwnershipFilters() {
+    return Object.keys(ownershipFilter).map((filter) => {
```
Comment (ash3rz) [inline src/components/analyses/Header.js:32]: Same possible refactor here using `Object.values`

File: src/components/analyses/AnalysesNavigation.js [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+function BatchFilter(props) {
+    const { baseId, name, classes, onClearBatch } = props;
+    return (
+        <>
+            <Tooltip
+                title={getMessage("viewingBatch", {
+                    values: { name: name },
+                })}
+                id={build(baseId, ids.BATCH_FILTER, name)}
```
Comment (ash3rz) [inline src/components/analyses/AnalysesNavigation.js:72]: This should probably also have an `aria-label`

File: src/components/analyses/AnalysesNavigation.js [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+                    orientation="vertical"
+                    flexItem
+                />
+            </Hidden>
+            <Hidden smUp>
+                <Tooltip title={getMessage("viewAll")}>
+                    <IconButton
+                        id={build(baseId, ids.CLEAR_BATCH_FILTER, name)}
+                        size="small"
+                        onClick={onClearBatch}
```
Comment (ash3rz) [inline src/components/analyses/AnalysesNavigation.js:94]: Either the `Tooltip` or the `IconButton` should have `aria-label`

File: src/components/analyses/ids.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
+export default {
+    CHECKBOX: ".checkbox",
```
Comment (ash3rz) [inline src/components/analyses/ids.js:2]: Just picking a random spot - these ids don't need the `.` as long as you're using `build` for your ids.

File: src/components/analyses/listing/TableView.js [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+    const NameLink = () => {
+        return (
+            <Tooltip
+                title={formatMessage(intl, "goOutputFolderOf") + " " + name}
+            >
+                <Link
+                    onClick={() => handleGoToOutputFolder(analysis)}
+                    color="primary"
+                >
+                    <span style={{ paddingLeft: 8, cursor: "pointer" }}>
```
Comment (ash3rz) [inline src/components/analyses/listing/TableView.js:62]: This should probably be `theme.spacing(1)`

File: src/components/analyses/listing/TableView.js [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+                </Link>
+            </Tooltip>
+        );
+    };
+
+    if (isBatch) {
+        return (
+            <>
+                <Tooltip title={getMessage("htDetails")}>
+                    <IconButton
```
Comment (ash3rz) [inline src/components/analyses/listing/TableView.js:74]: Either the `Tooltip` or `IconButton` should have `aria-label`

File: src/components/analyses/listing/TableView.js [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+        (status === analysisStatus.SUBMITTED ||
+            status === analysisStatus.RUNNING) &&
+        interactiveUrls &&
+        interactiveUrls.length > 0
+    ) {
+        return (
+            <>
+                <Tooltip title={getMessage("goToVice")}>
+                    <IconButton
+                        onClick={() =>
```
Comment (ash3rz) [inline src/components/analyses/listing/TableView.js:95]: Either the `Tooltip` or `IconButton` should have `aria-label`

File: src/components/analyses/listing/TableView.js [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+    const handleInteractiveUrlClick = props.handleInteractiveUrlClick;
+    const handleGoToOutputFolder = props.handleGoToOutputFolder;
+    const handleBatchIconClick = props.handleBatchIconClick;
+    const intl = props.intl;
+    const baseId = props.baseId;
+
+    const NameLink = () => {
+        return (
+            <Tooltip
+                title={formatMessage(intl, "goOutputFolderOf") + " " + name}
```
Comment (ash3rz) [inline src/components/analyses/listing/TableView.js:56]: Just a reminder that values can be passed to `formatMessage`
```
formatMessage(intl, "goOutputFolderOf", {name: name})
```

File: src/components/analyses/listing/TableView.js [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+                    rowsInPage={listing?.analyses ? listing.analyses.length : 0}
+                    order={order}
+                    orderBy={orderBy}
+                    columnData={columnData}
+                    onRequestSort={handleRequestSort}
+                    onSelectAllClick={handleSelectAllClick}
+                />
+                {loading && (
+                    <TableLoading
+                        numColumns={6}
```
Comment (ash3rz) [inline src/components/analyses/listing/TableView.js:236]: This could be `columnData.length`

File: src/components/analyses/listing/TableView.js [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+                                        aria-checked={isSelected}
+                                        tabIndex={-1}
+                                        selected={isSelected}
+                                        hover
+                                        key={id}
+                                        id={rowId}
+                                    >
+                                        <TableCell padding="checkbox">
+                                            <DECheckbox
+                                                id={build(rowId + ids.CHECKBOX)}
```
Comment (ash3rz) [inline src/components/analyses/listing/TableView.js:281]: Just picking a spot - our pattern has been to do `build(rowId, ids.CHECKBOX)`

File: src/components/analyses/messages.js [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+            " analysis changes to Completed or Failed.",
+        jobLogsUnavailableHeading: "Logs not available",
+        goToVice: "Go to analysis",
+        extendTime: "Extend time limit",
+        extendTimeLimitMessage:
+            "This analysis is scheduled to be terminated at {timeLimit}. Do you" +
+            " wish" +
+            " to extend the time limit?",
+        analysesExecDeleteWarning:
+            "This will remove the selected analyses and the parameters information associated with those analyses. Outputs can still be viewed in the Data window within the folder created by these analyses.",
```
Comment (ash3rz) [inline src/components/analyses/messages.js:72]: No more windows! 🎉 
Maybe change to `...still be viewed in the Data view within the...`

File: src/serviceFacade/analysesServiceFacade.js [hunk 14]
```diff
@@ -0,0 +1,11 @@
+import callApi from "../common/callApi";
+
+function getAnalyses(key, { rowsPerPage, orderBy, order, page, filter }) {
```
Comment (ash3rz) [inline src/serviceFacade/analysesServiceFacade.js:3]: I think we're chopping off `ServiceFacade` from these filenames now, so this file should just be `analyses.js`

File: stories/analyses/Listing.stories.js [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+import { mockAxios } from "../axiosMock";
+import { listing } from "./AnalysesMocks";
+import { UserProfileProvider } from "../../src/contexts/userProfile";
+export default {
+    title: "Analyses",
+};
+
+function ListingTest(props) {
+    return (
+        <UserProfileProvider>
```
Comment (ash3rz) [inline stories/analyses/Listing.stories.js:12]: You shouldn't need this `UserProfileProvider` wrapper anymore.  It's in the storybook config (because I kept forgetting to add it to my stories and it was driving me crazy lol)

File: src/components/analyses/listing/TableView.js [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+    const handleInteractiveUrlClick = props.handleInteractiveUrlClick;
+    const handleGoToOutputFolder = props.handleGoToOutputFolder;
+    const handleBatchIconClick = props.handleBatchIconClick;
+    const intl = props.intl;
+    const baseId = props.baseId;
+
+    const NameLink = () => {
+        return (
+            <Tooltip
+                title={formatMessage(intl, "goOutputFolderOf", { name: name })}
```
Comment (psarando) [inline src/components/analyses/listing/TableView.js:65]: I think this should also be used in an `aria-label`.

File: src/components/analyses/listing/TableView.js [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+                </Link>
+            </Tooltip>
+        );
+    };
+
+    if (isBatch) {
+        return (
+            <>
+                <NameLink />
+                <Tooltip title={getMessage("htDetails")}>
```
Comment (psarando) [inline src/components/analyses/listing/TableView.js:82]: I think the `Tooltip` needs the `aria-label`.

File: src/components/analyses/listing/TableView.js [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+    } else if (
+        (status === analysisStatus.SUBMITTED ||
+            status === analysisStatus.RUNNING) &&
+        interactiveUrls &&
+        interactiveUrls.length > 0
+    ) {
+        return (
+            <>
+                <NameLink />
+                <Tooltip title={getMessage("goToVice")}>
```
Comment (psarando) [inline src/components/analyses/listing/TableView.js:103]: I think the `Tooltip` needs the `aria-label`.

File: src/components/analyses/listing/TableView.js [hunk 19]
```diff
@@ -0,0 +1,366 @@
...
+    return <Link component="button">{name}</Link>;
+}
+
+function Status(props) {
+    const { analysis, username, baseId, analysisUser, intl } = props;
+    const allowTimeExtn =
+        analysis.interactive_urls &&
+        analysis.interactive_urls.length > 0 &&
+        analysis.status === analysisStatus.RUNNING &&
+        username === analysisUser;
```
Comment (psarando) [inline src/components/analyses/listing/TableView.js:134]: 👍

File: src/components/analyses/listing/TableView.js [hunk 20]
```diff
@@ -0,0 +1,366 @@
...
+    {
+        id: ids.APP,
+        name: "App",
+        numeric: false,
+        enableSorting: false,
+        key: "app",
+    },
+    {
+        id: ids.START_DATE,
+        name: "Start Date",
```
Comment (psarando) [inline src/components/analyses/listing/TableView.js:179]: I think these column names should come from i18n messages.

File: src/components/analyses/listing/TableView.js [hunk 21]
```diff
@@ -0,0 +1,366 @@
...
+                id={tableId}
+                stickyHeader={true}
+                size="small"
+                aria-label={formatMessage(intl, "ariaTableListing")}
+            >
+                <EnhancedTableHead
+                    baseId={baseId}
+                    selectable={true}
+                    numSelected={selected.length}
+                    rowsInPage={listing?.analyses ? listing.analyses.length : 0}
```
Comment (psarando) [inline src/components/analyses/listing/TableView.js:240]: This could be condensed (it's a little tricky to read as-is):
```suggestion
                    rowsInPage={listing?.analyses?.length || 0}
```

File: src/components/analyses/listing/TableView.js [hunk 22]
```diff
@@ -0,0 +1,366 @@
...
+            <>
+                <NameLink />
+                <Tooltip title={getMessage("htDetails")}>
+                    <IconButton
+                        size="small"
+                        onClick={() => handleBatchIconClick(analysis)}
+                        id={build(baseId, ids.ICONS.BATCH)}
+                        aria-label={formatMessage(intl, "htDetails")}
+                    >
+                        <UnfoldMoreIcon />
```
Comment (psarando) [inline src/components/analyses/listing/TableView.js:89]: 👍

**Comment by sri2k1us** [comment]
> I also agree with @ash3rz that we should discuss action-icon placement in the table rows.

Sounds good. I will leave them in place for now. But we certainly take that up in one of the UX meetings.

File: src/components/analyses/listing/Listing.js [hunk 23]
```diff
@@ -0,0 +1,290 @@
...
+        let newSelected = [...selected];
+        analysesIds.forEach((analysisId) => {
+            const selectedIndex = selected.indexOf(analysisId);
+            if (selectedIndex === -1) {
+                newSelected.push(analysisId);
+            }
+        });
+
+        setSelected(newSelected);
+    };
```
[the remark this answers is not in the document]
Comment (sri2k1us) [reply src/components/analyses/listing/Listing.js:143]: Sorry! Somehow I missed it when addressing the PR comments in #77. Let me update the code in Apps as well.

File: src/components/analyses/ids.js [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+    APP: "app",
+    START_DATE: "start_date",
+    END_DATE: "end_date",
+    STATUS: "status",
+    TOOLBAR: "menuBar",
+    PARAMETERS: "parameters",
+    TYPE: "type",
+    VALUE: "value",
+    JOB_ID: "jobId",
+    ANALYSES_VIEW: ".analysesView",
```
[the remark this answers is not in the document]
Comment (sri2k1us) [reply src/components/analyses/ids.js:14]: I Simply copy-pasted all the `ids` and `messages` from DE 😉 . I was hoping to fix them as when I put them to use. I will go ahead remove all the `.`  from the `ids.

File: stories/analyses/Listing.stories.js (hunk 15 above)
Comment (sri2k1us) [reply stories/analyses/Listing.stories.js:12]: Hmm jest test fails if remove the `<UserProfileProvider>`.

Analyses Listing view

    useUserProfile must be used within a UserProfileProvider

      12 |     const context = React.useContext(UserProfileContext);
      13 |     if (!context) {
    > 14 |         throw new Error(
         |               ^
      15 |             `useUserProfile must be used within a UserProfileProvider`
      16 |         );
      17 |     }

      at useUserProfile (src/contexts/userProfile.js:14:15)
      at Listing (src/components/analyses/listing/Listing.js:63:27)

File: src/components/analyses/model/ownershipFilter.js [hunk 25]
```diff
@@ -0,0 +1,7 @@
+const ownershipFilter = {
+    all: "All",
+    mine: "Only my analyses",
+    theirs: "Analyses shared with me",
```
[the remark this answers is not in the document]
Comment (sri2k1us) [reply src/components/analyses/model/ownershipFilter.js:4]: These values are sent to the service and have to be in english ? 🤔

File: stories/analyses/Listing.stories.js (hunk 15 above)
Comment (sri2k1us) [reply stories/analyses/Listing.stories.js:12]: I think `data` jest passes because it is only testing table view. But am testing the whole listing in `analyses` jest. I will go ahead and modify the `analyses` jest to test only table view.

**Comment by sri2k1us** [comment]
@psarando Can I merge this PR ? (after resolving conflict)

File: src/components/analyses/ids.js (hunk 24 above)
[the remark this answers is not in the document]
Comment (psarando) [reply src/components/analyses/ids.js:14]: Yep, I could tell they were copied from the GWT code 😉

File: src/components/analyses/listing/Listing.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (psarando) [reply src/components/analyses/listing/Listing.js]: No problem! Some of these big PRs get a lot of comments to address 😅

File: src/components/analyses/listing/Listing.js [hunk 26]
```diff
@@ -159,13 +152,17 @@ function Listing(props) {
     };
 
     const rangeSelect = (start, end, targetId) => {
-        const rangeIds = [];
-        for (let i = start; i <= end; i++) {
-            rangeIds.push(data?.analyses[i].id);
+        // when a user first click on a row with shift key pressed,
+        // start is -1 (which is lastSelectIndex) and
+        // results in an error (data.analyses[-1].id)
+        if (start > -1) {
```
Comment (psarando) [inline src/components/analyses/listing/Listing.js:158]: 👍

File: src/components/analyses/listing/TableView.js [hunk 27]
```diff
@@ -62,12 +62,22 @@ function AnalysisName(props) {
...
                 <Link
+                    id={build(baseId, ids.ANALYSIS_NAME_CELL)}
                     onClick={() => handleGoToOutputFolder(analysis)}
                     color="primary"
                     component="button"
+                    aria-describedby={build(
+                        baseId,
+                        ids.ANALYSIS_NAME_CELL,
+                        ids.TOOLTIP
+                    )}
```
Comment (psarando) [inline src/components/analyses/listing/TableView.js:80]: Ah, my comment wasn't meant to instruct you to add an `aria-describedby` for Tooltips.

What I meant was that MUI does it automatically for us, with their `Tooltip` component.
So if MUI's `Tooltip` is adding an `aria-describedby` automatically to this `Link`, then it makes sense to me that the `Link` doesn't need an extra `aria-label`, especially if we follow MUI's examples and add the `aria-label` to the `Tooltip` (as you have done above).

In other words, we can leave the `aria-label` on the `Tooltip`, but we don't need `aria-describedby` for children of Tooltips, since MUI adds them automatically.

File: src/components/analyses/listing/TableView.js [hunk 28]
```diff
@@ -152,45 +190,45 @@ function Status(props) {
     );
 }
 
-const columnData = [
+const columnData = (intl) => [
```
Comment (psarando) [inline src/components/analyses/listing/TableView.js:193]: 👍

File: src/components/analyses/model/ownershipFilter.js (hunk 25 above)
[the remark this answers is not in the document]
Comment (psarando) [reply src/components/analyses/model/ownershipFilter.js:4]: Actually, it looks like these strings are only used for display in the `Autocomplete` component for the permission filter. In the `Listing` component's `useEffect`, there is a `switch` that checks these strings but uses a `Filters` constant that is sent to the service:
https://github.com/cyverse-de/sonora/blob/7ae319e351c55e6d8d6073009f6fb7993dc67a54/src/components/analyses/listing/Listing.js#L29-L37

File: src/components/analyses/listing/Listing.js [hunk 29]
```diff
@@ -0,0 +1,287 @@
...
+        } else {
+            select([analysisId]);
+        }
+    };
+
+    const rangeSelect = (start, end, targetId) => {
+        // when a user first click on a row with shift key pressed,
+        // start is -1 (which is lastSelectIndex) and
+        // results in an error (data.analyses[-1].id)
+        if (start > -1) {
```
Comment (psarando) [inline src/components/analyses/listing/Listing.js:158]: 👍

**Comment by sri2k1us** [comment]
thanks for reviews.
